From 727399da3c5fa495c13e91d05383f709068f1287 Mon Sep 17 00:00:00 2001
From: Terminator-J <terminator_j@hotbutteredmonkey.com>
Date: Wed, 15 Dec 2021 14:38:37 -0800
Subject: [PATCH] sdm845-common: packages: Include prebuilt OnePlus Camera

Includes adaptations of Vachounet's "set client package name"
and AnierinB's "Add OnePlus Camera support!" commits.

Co-authored-by: DennySPB <dennyspb@gmail.com>
Co-authored-by: Vachounet <vachounet@live.fr>
Co-authored-by: AnierinB <anierin@evolution-x.org>
---
 BoardConfigCommon.mk                                  |  5 +++++
 common.mk                                             | 10 ++++++++++
 .../android.hardware.camera.provider@2.4-service.rc   |  8 ++++++++
 lineage.dependencies                                  |  4 ++++
 rootdir/Android.mk                                    |  8 ++++++++
 rootdir/etc/init.onepluscamera.rc                     |  5 +++++
 sepolicy/private/cameraserver.te                      |  5 +++++
 sepolicy/private/file.te                              |  2 ++
 sepolicy/private/file_contexts                        |  3 ++-
 sepolicy/private/hal_cameraHIDL_default.te            |  3 +++
 sepolicy/private/platform_app.te                      |  2 ++
 sepolicy/private/system_server.te                     | 11 +++++++++++
 system.prop                                           |  9 ++++++++-
 13 files changed, 73 insertions(+), 2 deletions(-)
 create mode 100644 configs/vendor_overlay/android.hardware.camera.provider@2.4-service.rc
 create mode 100644 rootdir/etc/init.onepluscamera.rc
 create mode 100644 sepolicy/private/cameraserver.te
 create mode 100644 sepolicy/private/hal_cameraHIDL_default.te
 create mode 100644 sepolicy/private/platform_app.te
 create mode 100644 sepolicy/private/system_server.te

diff --git a/BoardConfigCommon.mk b/BoardConfigCommon.mk
index b5a6028..18882ab 100644
--- a/BoardConfigCommon.mk
+++ b/BoardConfigCommon.mk
@@ -81,6 +81,11 @@ ifeq ($(HOST_OS),linux)
   endif
 endif

+# Camera
+TARGET_CAMERA_NEEDS_CLIENT_INFO := true
+TARGET_USES_QTI_CAMERA_DEVICE := true
+USE_DEVICE_SPECIFIC_CAMERA := true
+
 # Display
 TARGET_USES_HWC2 := true

diff --git a/common.mk b/common.mk
index 24a0960..2b2bd29 100644
--- a/common.mk
+++ b/common.mk
@@ -28,6 +28,9 @@ $(call inherit-product, frameworks/native/build/phone-xhdpi-6144-dalvik-heap.mk)
 # Get non-open-source specific aspects
 $(call inherit-product, vendor/oneplus/sdm845-common/sdm845-common-vendor.mk)

+LOCAL_STATIC_JAVA_LIBRARIES := \
+    vendor.oneplus.hardware.camera-V1.0-java
+
 # Overlays
 DEVICE_PACKAGE_OVERLAYS += \
     $(LOCAL_PATH)/overlay \
@@ -102,6 +105,7 @@ PRODUCT_PACKAGES += \

 # Common init scripts
 PRODUCT_PACKAGES += \
+    init.onepluscamera.rc \
     init.qcom.rc \
     init.recovery.qcom.rc \
     ueventd.qcom.rc
@@ -167,6 +171,12 @@ PRODUCT_PACKAGES += \
 PRODUCT_BOOT_JARS += \
     oneplus-fwk

+# OnePlus Camera
+$(call inherit-product-if-exists, packages/apps/OnePlusCamera/config.mk)
+
+PRODUCT_COPY_FILES += \
+    $(LOCAL_PATH)/configs/vendor_overlay/android.hardware.camera.provider@2.4-service.rc:$(TARGET_COPY_OUT_PRODUCT)/vendor_overlay/$(PRODUCT_TARGET_VNDK_VERSION)/etc/init/android.hardware.camera.provider@2.4-service.rc
+
 # Power
 PRODUCT_PACKAGES += \
     power.qcom
diff --git a/configs/vendor_overlay/android.hardware.camera.provider@2.4-service.rc b/configs/vendor_overlay/android.hardware.camera.provider@2.4-service.rc
new file mode 100644
index 0000000..f7ac9f8
--- /dev/null
+++ b/configs/vendor_overlay/android.hardware.camera.provider@2.4-service.rc
@@ -0,0 +1,8 @@
+service vendor.camera-provider-2-4 /vendor/bin/hw/android.hardware.camera.provider@2.4-service
+    interface android.hardware.camera.provider@2.4::ICameraProvider legacy/0
+    class hal
+    user cameraserver
+    group audio camera input drmrpc
+    ioprio rt 4
+    capabilities SYS_NICE
+    task_profiles CameraServiceCapacity MaxPerformance
diff --git a/lineage.dependencies b/lineage.dependencies
index a445e08..278430f 100644
--- a/lineage.dependencies
+++ b/lineage.dependencies
@@ -6,5 +6,9 @@
   {
     "repository": "android_kernel_oneplus_sdm845",
     "target_path": "kernel/oneplus/sdm845"
+  },
+  {
+    "repository": "android_packages_apps_OnePlusCamera",
+    "target_path": "packages/apps/OnePlusCamera"
   }
 ]
diff --git a/rootdir/Android.mk b/rootdir/Android.mk
index 7eb4e7a..b503807 100644
--- a/rootdir/Android.mk
+++ b/rootdir/Android.mk
@@ -1,5 +1,13 @@
 LOCAL_PATH := $(call my-dir)

+include $(CLEAR_VARS)
+LOCAL_MODULE       := init.onepluscamera.rc
+LOCAL_MODULE_TAGS  := optional
+LOCAL_MODULE_CLASS := ETC
+LOCAL_SRC_FILES    := etc/init.onepluscamera.rc
+LOCAL_MODULE_PATH  := $(TARGET_OUT_ETC)/init
+include $(BUILD_PREBUILT)
+
 include $(CLEAR_VARS)
 LOCAL_MODULE       := init.qcom.rc
 LOCAL_MODULE_TAGS  := optional
diff --git a/rootdir/etc/init.onepluscamera.rc b/rootdir/etc/init.onepluscamera.rc
new file mode 100644
index 0000000..a8e4e1c
--- /dev/null
+++ b/rootdir/etc/init.onepluscamera.rc
@@ -0,0 +1,5 @@
+on boot
+    mkdir /data/misc/lineage 0770 cameraserver audio
+    write /data/misc/lineage/client_package_name 0
+    chown cameraserver audio /data/misc/lineage/client_package_name
+    chmod 0644 /data/misc/lineage/client_package_name
diff --git a/sepolicy/private/cameraserver.te b/sepolicy/private/cameraserver.te
new file mode 100644
index 0000000..2d2f926
--- /dev/null
+++ b/sepolicy/private/cameraserver.te
@@ -0,0 +1,5 @@
+allow cameraserver lineage_data_file:dir write;
+allow cameraserver lineage_data_file:file rw_file_perms;
+
+allow cameraserver init:unix_stream_socket { connectto };
+allow cameraserver property_socket:sock_file rw_file_perms;
diff --git a/sepolicy/private/file.te b/sepolicy/private/file.te
index 6354219..a98a072 100644
--- a/sepolicy/private/file.te
+++ b/sepolicy/private/file.te
@@ -13,7 +13,9 @@ type sysfs_oem, sysfs_type, fs_type;
 type sysfs_ssr, sysfs_type, fs_type;
 type sysfs_ssr_toggle,  sysfs_type, fs_type;
 type sysfs_usb_supply, sysfs_type, fs_type;
+type vendor_sysfs_battery_supply, sysfs_type, fs_type;
 type vendor_sysfs_usb_supply, sysfs_type, fs_type;

 # data
+type lineage_data_file, file_type, data_file_type, core_data_file_type;
 type display_misc_file, file_type, data_file_type, core_data_file_type;
diff --git a/sepolicy/private/file_contexts b/sepolicy/private/file_contexts
index be8c841..282877a 100644
--- a/sepolicy/private/file_contexts
+++ b/sepolicy/private/file_contexts
@@ -1,5 +1,6 @@
 # Data files
-/data/misc/display(/.*)?      u:object_r:display_misc_file:s0
+/data/misc/display(/.*)?                  u:object_r:display_misc_file:s0
+/data/misc/lineage/client_package_name    u:object_r:lineage_data_file:s0

 # Files in rootfs
 /op1(/.*)?       u:object_r:op1_file:s0
diff --git a/sepolicy/private/hal_cameraHIDL_default.te b/sepolicy/private/hal_cameraHIDL_default.te
new file mode 100644
index 0000000..fe8ec74
--- /dev/null
+++ b/sepolicy/private/hal_cameraHIDL_default.te
@@ -0,0 +1,3 @@
+type hal_cameraHIDL_default, domain, mlstrustedsubject;
+binder_call(cameraserver, hal_cameraHIDL_default)
+binder_call(system_server, hal_cameraHIDL_default)
diff --git a/sepolicy/private/platform_app.te b/sepolicy/private/platform_app.te
new file mode 100644
index 0000000..ffceb13
--- /dev/null
+++ b/sepolicy/private/platform_app.te
@@ -0,0 +1,2 @@
+allow platform_app vendor_sysfs_battery_supply:dir r_dir_perms;
+allow platform_app vendor_sysfs_battery_supply:file r_file_perms;
diff --git a/sepolicy/private/system_server.te b/sepolicy/private/system_server.te
new file mode 100644
index 0000000..b6be57d
--- /dev/null
+++ b/sepolicy/private/system_server.te
@@ -0,0 +1,11 @@
+allow system_server lineage_data_file:dir search;
+allow system_server lineage_data_file:file r_file_perms;
+
+type hal_cameraHIDL_hwservice, hwservice_manager_type;
+add_hwservice(system_server, hal_cameraHIDL_hwservice)
+allow system_server hal_cameraHIDL_hwservice:hwservice_manager { find };
+allow system_server hal_cameraHIDL_default:binder { call };
+
+get_prop(system_server, vendor_camera_prop)
+
+allow system_server vendor_sysfs_battery_supply:file rw_file_perms;
diff --git a/system.prop b/system.prop
index a9525e9..bec03ab 100644
--- a/system.prop
+++ b/system.prop
@@ -11,11 +11,18 @@ ro.kernel.ebpf.supported=true

 # Camera
 camera.disable_zsl_mode=true
-vendor.camera.aux.packagelist=com.oneplus.camera,com.oneplus.camera.service,org.codeaurora.snapcam
+persist.camera.assert.panic=true
+persist.vendor.camera.privapp.list=com.oneplus.camera
+ro.opcamera.support=true
+vendor.camera.aux.packagelist=com.oneplus.camera,org.codeaurora.snapcam
+vendor.camera.skip_unconfigure.packagelist=com.oneplus.camera

 # Charger
 ro.charger.disable_init_blank=true

+# Codec
+debug.stagefright.ccodec=1
+
 # Display
 debug.gralloc.gfx_ubwc_disable=0
 debug.sf.enable_hwc_vds=1
--
2.25.1

